plugins/usi-dock/fu-usi-dock-mcu-device.c

```diff
@@ -0,0 +1,777 @@
+		if (buf[6] == FIRMWARE_IDX_AUDIO) {
+			g_debug("got a quick jump at audio updates");
+		} else {
+			g_debug("got a ignored report");
+			return FALSE;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_usi_dock_mcu_device_tx(FuUsiDockMcuDevice *self,
+			  guint8 tag2,
+			  const guint8 *inbuf,
+			  gsize inbufsz,
+			  GError **error)
+{
+	guint8 buf[64] = {USB_HID_REPORT_ID2,
+			  0x3 + inbufsz /* length */,
+			  0xFE /* tag1: MCU defined */,
+			  0xFF /* tag2: MCU defined */};
+	buf[63] = tag2;
+	if (inbuf != NULL) {
+		if (!fu_memcpy_safe(buf,
+				    sizeof(buf),
+				    0x4, /* dst */
+				    inbuf,
+				    inbufsz,
+				    0x0, /* src */
+				    inbufsz,
+				    error))
+			return FALSE;
+	}
+
+	/* special cases */
+	if (buf[5] == USBUID_ISP_INTERNAL_FW_CMD_UPDATE_FW) {
+		buf[6] = 0xFF;
+	}
+
+	return fu_hid_device_set_report(FU_HID_DEVICE(self),
+					USB_HID_REPORT_ID2,
+					buf,
+					sizeof(buf),
+					FU_USI_DOCK_MCU_DEVICE_TIMEOUT,
+					FU_HID_DEVICE_FLAG_NONE,
+					error);
+}
+
+static gboolean
+fu_usi_dock_mcu_device_rx(FuUsiDockMcuDevice *self,
+			  guint8 cmd,
+			  guint8 *outbuf,
+			  gsize outbufsz,
+			  GError **error)
+{
+	guint8 buf[64] = {0};
+
+	if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+				      USB_HID_REPORT_ID2,
+				      buf,
+				      sizeof(buf),
+				      FU_USI_DOCK_MCU_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		return FALSE;
+	}
+	if (buf[0] != USB_HID_REPORT_ID2) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid ID, expected 0x%02x, got 0x%02x",
+			    USB_HID_REPORT_ID2,
+			    buf[0]);
+		return FALSE;
+	}
+	if (buf[2] != 0xFE || buf[3] != 0xFF) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid tags, expected 0x%02x:0x%02x, got 0x%02x:0x%02x",
+			    0xFEu,
+			    0xFFu,
+			    buf[2],
+			    buf[3]);
+		return FALSE;
+	}
+
+	/* special case */
+	if (!fu_usi_dock_rx_filter(cmd, buf)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "no expected report received");
+		return FALSE;
+	}
+
+	if (outbuf != NULL) {
+		if (!fu_memcpy_safe(outbuf,
+				    outbufsz,
+				    0x0, /* dst */
+				    buf,
+				    sizeof(buf),
+				    0x5, /* src */
+				    outbufsz,
+				    error))
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_usi_dock_mcu_device_txrx(FuUsiDockMcuDevice *self,
+			    guint8 tag2,
+			    const guint8 *inbuf,
+			    gsize inbufsz,
+			    guint8 *outbuf,
+			    gsize outbufsz,
+			    GError **error)
+{
+	if (!fu_usi_dock_mcu_device_tx(self, tag2, inbuf, inbufsz, error))
+		return FALSE;
+	return fu_usi_dock_mcu_device_rx(self, USBUID_ISP_CMD_ALL, outbuf, outbufsz, error);
+}
+
+static gboolean
+fu_usi_dock_mcu_device_get_status(FuUsiDockMcuDevice *self, GError **error)
+{
+	guint8 buf[] = {USBUID_ISP_DEVICE_CMD_MCU_STATUS};
```

Comment:
buf is normally 64 bytes, why isn't this one?  Should it be set that way for safety?

Reply:
#4062